lib/CL/devices/level0/level0-driver.cc

```diff
@@ -0,0 +1,2239 @@
+                             mem_mapping_t *Map) {
+  char *DstImgPtr = static_cast<char *>(MemId->mem_ptr);
+
+  POCL_MSG_PRINT_LEVEL0("UNMAP IMAGE: %p FLAGS %zu\n", DstImgPtr,
+                        Map->map_flags);
+
+  /* for read mappings, don't copy anything */
+  if (Map->map_flags == CL_MAP_READ)
+    return;
+
+  assert(Map->host_ptr == (DstImgPtr + Map->offset));
+
+  writeImageRect(DstImage, MemId, DstImgPtr, nullptr, Map->origin, Map->region,
+                 Map->row_pitch, Map->slice_pitch, Map->offset);
+}
+
+/* fill image with pattern */
+void Level0Queue::fillImage(cl_mem Image, pocl_mem_identifier *MemId,
+                            const size_t *Origin, const size_t *Region,
+                            cl_uint4 OrigPixel, pixel_t FillPixel,
+                            size_t PixelSize) {
+  POCL_ABORT_UNIMPLEMENTED("fillImage");
+}
+
+void Level0Queue::svmMap(void *Ptr) { return; }
+
+void Level0Queue::svmUnmap(void *Ptr) { return; }
+
+void Level0Queue::svmCopy(void *DstPtr, const void *SrcPtr, size_t Size) {
+  POCL_MSG_PRINT_LEVEL0("SVM COPY | SRC %p | DST %p | SIZE %zu\n", SrcPtr,
+                        DstPtr, Size);
+
+  ze_result_t res = zeCommandListAppendMemoryCopy(CmdListH, DstPtr, SrcPtr,
+                                                  Size, NULL, 0, NULL);
+  LEVEL0_CHECK_ABORT(res);
+}
+
+void Level0Queue::svmFill(void *DstPtr, size_t Size, void *Pattern,
+                          size_t PatternSize) {
+  POCL_MSG_PRINT_LEVEL0("SVM FILL | PTR %p | SIZE %zu | PAT SIZE %zu\n", DstPtr,
+                        Size, PatternSize);
+  ze_result_t Res = zeCommandListAppendMemoryFill(
+      CmdListH, DstPtr, Pattern, PatternSize, Size, nullptr, 0, nullptr);
+  LEVEL0_CHECK_ABORT(Res);
+}
+
+bool Level0Queue::setupKernelArgs(ze_module_handle_t ModuleH,
+                                  ze_kernel_handle_t KernelH, cl_device_id Dev,
+                                  unsigned DeviceI, _cl_command_run *RunCmd) {
+  cl_kernel Kernel = RunCmd->kernel;
+  struct pocl_argument *PoclArg = RunCmd->arguments;
+
+  /****************************************************************************************/
+
+  /* static locals are taken care of in ZE compiler */
+  assert(Kernel->meta->num_locals == 0);
+
+  cl_uint i;
+  ze_result_t Res;
+  for (i = 0; i < Kernel->meta->num_args; ++i) {
+    if (ARG_IS_LOCAL(Kernel->meta->arg_info[i])) {
+      assert(PoclArg[i].size > 0);
+      Res = zeKernelSetArgumentValue(KernelH, i, PoclArg[i].size, NULL);
+      LEVEL0_CHECK_ABORT(Res);
+
+    } else if (Kernel->meta->arg_info[i].type == POCL_ARG_TYPE_POINTER) {
+      assert(PoclArg[i].size == sizeof(void *));
+
+      if (PoclArg[i].value == NULL) {
+        Res = zeKernelSetArgumentValue(KernelH, i, sizeof(void *), nullptr);
+      } else if (PoclArg[i].is_svm) {
+        void *MemPtr = PoclArg[i].value;
+        Res = zeKernelSetArgumentValue(KernelH, i, sizeof(void *), &MemPtr);
```

Comment:
Do we need to set some flags for the shared buffers (IIRC)?

Reply:
We currently use `zeMemAllocShared` for all buffers, and the driver should take care of necessary migrations. But i added a TODO note in the driver documentation.